walk/jujutsu.go

```diff
@@ -0,0 +1,171 @@
+package walk
+
+import (
+	"bufio"
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strconv"
+
+	"github.com/charmbracelet/log"
+	"github.com/numtide/treefmt/v2/jujutsu"
+	"github.com/numtide/treefmt/v2/stats"
+	"golang.org/x/sync/errgroup"
+)
+
+type JujutsuReader struct {
+	root string
+	path string
+
+	log   *log.Logger
+	stats *stats.Stats
+
+	eg      *errgroup.Group
+	scanner *bufio.Scanner
+}
+
+func (j *JujutsuReader) Read(ctx context.Context, files []*File) (n int, err error) {
+	// ensure we record how many files we traversed
+	defer func() {
+		j.stats.Add(stats.Traversed, n)
+	}()
+
+	if j.scanner == nil {
+		// create a pipe to capture the command output
+		r, w := io.Pipe()
+
+		// create a command which will execute from root
+		// --ignore-working-copy: Don't snapshot the working copy, and don't update it. This prevents that the user has to
+		// enter a password for singning the commit. New files also won't be added to the index and not displayed in the
+		// output.
+		// Add the subpath as a fileset displaying only files matching this prefix. If
+		// the subpath is empty ignore it since it interferese with command
+		args := []string{"file", "list", "--ignore-working-copy"}
+		if j.path != "" {
+			args = append(args, j.path)
+		}
+
+		cmd := exec.Command("jj", args...)
+		cmd.Stdout = w
+
+		// execute the command in the background
+		j.eg.Go(func() error {
+			return w.CloseWithError(cmd.Run())
+		})
+
+		// create a new scanner for reading the output
+		j.scanner = bufio.NewScanner(r)
+	}
+
+	nextLine := func() (string, error) {
+		line := j.scanner.Text()
+
+		if len(line) == 0 || line[0] != '"' {
```

Comment:
> `line[0] != '"'`

This needs some explanation. What are we skipping? Is this really safe? Is there a more structured (json?) output we could consume instead?

Reply:
This function returns lines without quotes. If a line is empty or does not start with a quote, it is returned immediately. Lines that are quoted are unquoted on line 71 before being returned. However, I'm not sure why this check is necessary, as neither `git ls-files` nor `jj file list` appear to return quoted paths. I have this in here because `walk/git.go` does too.

Reply:
I observed that `git ls-files` in this repository returns:

```
...
"test/examples/emoji \360\237\225\260\357\270\217/README.md"
...
```

Whereas `jj file list` outputs:

```
...
test/examples/emoji 🕰️/README.md
...
```

Therefore, it seems that unquoting can be removed for Jujutsu.